services/provider/server/consumer.go

```diff
@@ -23,68 +25,56 @@ type ocsConsumerManager struct {
 	mutex     sync.RWMutex
 }
 
-func newConsumerManager(ctx context.Context, cl client.Client, namespace string) (*ocsConsumerManager, error) {
-	consumers := &ocsv1alpha1.StorageConsumerList{}
-	err := cl.List(ctx, consumers, client.InNamespace(namespace))
-	if err != nil {
-		return nil, fmt.Errorf("failed to list storage consumers. %v", err)
-	}
-
-	nameByUID := map[types.UID]string{}
-
-	for _, consumer := range consumers.Items {
-		nameByUID[consumer.UID] = consumer.Name
-	}
+func newConsumerManager(ctx context.Context, cl client.Client, namespace string, cache cache.Cache) (*ocsConsumerManager, error) {
 
-	return &ocsConsumerManager{
+	consumerManager := &ocsConsumerManager{
 		client:    cl,
 		namespace: namespace,
-		nameByUID: nameByUID,
-	}, nil
-}
-
-// Delete deletes the storageConsumer resource using UID and updates the consumer cache
-func (c *ocsConsumerManager) Delete(ctx context.Context, id string) error {
-	uid := types.UID(id)
-	c.mutex.RLock()
-	consumerName, ok := c.nameByUID[uid]
-	if !ok {
-		klog.Warningf("no storageConsumer found with UID %q", id)
-		c.mutex.RUnlock()
-		return nil
+		nameByUID: map[types.UID]string{},
 	}
-	c.mutex.RUnlock()
 
-	consumerObj := &ocsv1alpha1.StorageConsumer{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      consumerName,
-			Namespace: c.namespace,
+	consumer := &metav1.PartialObjectMetadata{}
+	consumer.SetGroupVersionKind(ocsv1alpha1.GroupVersion.WithKind("StorageConsumer"))
+
+	informer, err := cache.GetInformer(ctx, consumer)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get informer: %w", err)
+	}
+	if _, err = informer.AddEventHandler(clientgocache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj any) {
+			consumer, ok := obj.(*metav1.PartialObjectMetadata)
+			if !ok {
+				return
+			}
+			consumerManager.mutex.RLock()
+			consumerManager.nameByUID[consumer.GetUID()] = consumer.Name
+			consumerManager.mutex.RUnlock()
 		},
-	}
 
-	foregroundDelete := metav1.DeletePropagationForeground
-	deleteOption := client.DeleteOptions{
-		PropagationPolicy: &foregroundDelete,
-	}
-	if err := c.client.Delete(ctx, consumerObj, &deleteOption); err != nil {
-		if kerrors.IsNotFound(err) {
-			// update uidStore
-			c.mutex.Lock()
-			delete(c.nameByUID, uid)
-			c.mutex.Unlock()
-			klog.Warningf("storageConsumer %q not found.", consumerObj.Name)
-			return nil
-		}
-		return fmt.Errorf("failed to delete storageConsumer %q. %v", consumerName, err)
+		DeleteFunc: func(obj any) {
+			consumer, ok := obj.(*metav1.PartialObjectMetadata)
+			if !ok {
+				return
+			}
+			consumerManager.mutex.RLock()
+			delete(consumerManager.nameByUID, consumer.GetUID())
+			consumerManager.mutex.RUnlock()
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("failed to create informer: %w", err)
 	}
 
-	c.mutex.Lock()
-	delete(c.nameByUID, uid)
-	c.mutex.Unlock()
+	go func() {
+		if err := cache.Start(ctx); err != nil {
```

Comment:
what is the lifetime of this `ctx`, this shouldn't be less than server lifetime.

Reply:
the ctx here is the same as used for creating the NewProviderServer so it's lifetime shouldn't be lesser than server lifetime